bindings/rust/s2n-tls/src/client_hello.rs

```diff
@@ -0,0 +1,378 @@
+        connection::Connection,
+        error::{Error, ErrorType},
+        security,
+        testing::{poll_tls_pair, tls_pair},
+    };
+
+    use super::*;
+
+    /// Ensure memory is correctly managed in tests
+    /// tests invoked using the checkers::test macro have additional
+    /// memory sanity checks that occur
+    #[cfg(test)]
+    #[global_allocator]
+    static ALLOCATOR: checkers::Allocator = checkers::Allocator::system();
+
+    /// This function is a test fixture used a generate a valid ClientHello so
+    /// that we don't have to copy and paste the raw bytes for test fixtures
+    fn get_client_hello_bytes() -> Vec<u8> {
+        let config = crate::testing::config_builder(&security::DEFAULT_TLS13)
+            .unwrap()
+            .build()
+            .unwrap();
+        let pair = tls_pair(config);
+        let pair = poll_tls_pair(pair);
+        // this doesn't have the handshake header
+        let client_hello_message = pair
+            .server
+            .0
+            .connection()
+            .client_hello()
+            .unwrap()
+            .raw_message()
+            .unwrap();
+        // handshake header is {tag: u8, client_hello_length: u24}
+        let mut client_hello = vec![0; 4];
+        // As long as the client hello is small, no bit fiddling is required
+        assert!(client_hello_message.len() < u8::MAX as usize);
+        // tag for handshake header
+        client_hello[0] = 1;
+        client_hello[3] = client_hello_message.len() as u8;
+        client_hello.extend(client_hello_message.iter());
+        client_hello
+    }
+
+    // test that a fingerprint can successfully be calculated from ClientHellos
+    // returned from a connection
+    #[checkers::test]
+    fn io_fingerprint_test() {
+        let config = crate::testing::config_builder(&security::DEFAULT_TLS13)
+            .unwrap()
+            .build()
+            .unwrap();
+        let pair = crate::testing::tls_pair(config);
+
+        // client_hellos can not be accessed before the handshake
+        assert!(pair.client.0.connection().client_hello().is_err());
+        assert!(pair.server.0.connection().client_hello().is_err());
+
+        let pair = poll_tls_pair(pair);
+        let server_conn = pair.server.0.connection();
+        let client_conn = pair.server.0.connection();
+
+        let check_client_hello = |conn: &Connection| -> Result<(), Error> {
+            let client_hello = conn.client_hello().unwrap();
+            let (_hash, _string) = client_hello.fingerprint(FingerprintType::JA3)?;
+            Ok(())
+        };
+
+        assert!(check_client_hello(server_conn).is_ok());
+        assert!(check_client_hello(client_conn).is_ok());
+    }
+
+    fn known_test_case(
+        raw_client_hello: Vec<u8>,
+        expected_string: &str,
+        expected_hash_hex: &str,
+    ) -> Result<(), Error> {
+        let expected_hash: Vec<u8> = hex::decode(expected_hash_hex).unwrap();
+        crate::init::init();
```

Comment:
This should be in the `parse_client_hello` method so users don't have to call it (similar to the config and connection builders).